### PR TITLE
Added various security features issue #115

### PR DESCRIPTION
**Pull request by udaymittal7**
Added helmet for security http headers.
Added rate limiter to receive 100 requests per hour from a particular IP address.
Added mongo sanitize for data sanitization against noSQL query injection
Added xss for data sanitization against XSS.

Please review and if any changes required let me know.
Thank you.

For issue #115

### CODE REVIEW

**Review by ayan-biswas0412 — CHANGES_REQUESTED** [review]
your commit should not include the yarn.lock

**Comment by udaymittal7** [comment]
I am new to git. Can you tell me how to remove this commit from the pull request. Thank you.

**Comment by ayan-biswas0412** [comment]
> I am new to git. Can you tell me how to remove this commit from the pull request. Thank you.

REFER https://www.git-tower.com/learn/git/faq/undo-last-commit/

**Comment by udaymittal7** [comment]
I made the change. Please check. Thank you for the help.

**Comment by udaymittal7** [comment]
@ayan-biswas0412 I always get stuck when these conflicts occur. Please guide how to solve them.

File: backend/package.json [hunk 1]
```diff
@@ -14,9 +14,17 @@
     "cors": "^2.8.5",
     "dotenv": "^8.2.0",
     "express": "^4.17.1",
+<<<<<<< HEAD
```
Comment (singhayushh) [inline backend/package.json:17]: This commit has merge conflicts messages not removed which is a serious issue.
Kindly resolve the conflicts and remove the conflict messages.

**Review by singhayushh — CHANGES_REQUESTED** [review]

File: backend/package.json (hunk 1 above)
Comment (udaymittal7) [reply backend/package.json]: I don't know hot to resolve them. I have tried but they appear again and again. Please help.

**Review by singhayushh — APPROVED** [review]
Yeah it's alright now.

File: backend/yarn.lock [hunk 2]
```diff
@@ -272,14 +350,73 @@ debug@2.6.9:
...
   integrity sha512-bC7ElrdJaJnPbAP+1EotYvqZsb3ecl5wi6Bfi6BJTUcNowp6cvspg0jXznRTKDjm/E7AdgFBVeAPVMNcKGsHMA==
-  dependencies:
-    ms "2.0.0"
-
-debug@3.1.0:
-  version "3.1.0"
-  resolved "https://registry.yarnpkg.com/debug/-/debug-3.1.0.tgz#5bb5a0672628b64149566ba16819e61518c67261"
-  integrity sha512-OX8XqP7/1a9cqkxYw2yXss15f26NKWBpDXQd0/uK/KPqdQhxbPa994hnzjcE2VqQpDslf55723cKPUOGSmMY3g==
-  dependencies:
+>>>>>>> 9f3f9283bc5dfadd4b3254ea9c1cc91181621871
```
Comment (singhayushh) [inline backend/yarn.lock:353]: Here as well you have merge conflict messages unhandled
Kindly do not push this yarn.lock file in your commits

**Review by singhayushh — CHANGES_REQUESTED** [review]

File: backend/yarn.lock (hunk 2 above)
Comment (udaymittal7) [reply backend/yarn.lock:353]: Can you guide me how to remove it? Should I reset this commit using soft and then include every file except this yarn.lock file?

Comment (singhayushh) [reply backend/yarn.lock:353]: Yes you can do that. When you add extra dependencies, just make sure they are added in the package.json file.

Comment (udaymittal7) [reply backend/yarn.lock]: Please check now.

**Review by ayan-biswas0412 — CHANGES_REQUESTED** [review]
lots of merge conflicts there please resolve them

**Comment by udaymittal7** [comment]
I resolved the conflicts.

**Review by ayan-biswas0412 — CHANGES_REQUESTED** [review]
your code is failing ci checks

**Comment by udaymittal7** [comment]
> your code is failing ci checks

How to resolve them?

**Comment by ayan-biswas0412** [comment]
> > your code is failing ci checks
> 
> How to resolve them?
click on the CI to see what are failing

**Comment by udaymittal7** [comment]
It is just showing this check failing and nothing else.

**Comment by ayan-biswas0412** [comment]
> It is just showing this check failing and nothing else.

I have reinitiated ci jobs let's see what happens

**Comment by udaymittal7** [comment]
It is showing all checks are cleared now. FInally.

**Review by singhayushh — APPROVED** [review]

**Comment by ayan-biswas0412** [comment]
@singhayushh can you test this locally please ?

File: backend/package.json [hunk 3]
```diff
@@ -14,6 +14,11 @@
     "cors": "^2.8.5",
     "dotenv": "^8.2.0",
     "express": "^4.17.1",
+    "express-mongo-sanitize": "^2.0.2",
+    "express-rate-limit": "^5.2.6",
+    "helmet": "^4.4.1",
+    "mongoose": "^5.11.15",
```
Comment (singhayushh) [inline backend/package.json:20]: Hey, there's an error still. In the package.json file, line 20 and 23 are same + there is to be a comma at the end of line 21
Before issuing a PR next time, do run the app locally as these are some fatal bugs that totally break the server from even starting.

**Review by singhayushh — CHANGES_REQUESTED** [review]

File: backend/package.json (hunk 3 above)
Comment (udaymittal7) [reply backend/package.json]: I am sorry. I will check that in mind. I just made the changes though.

**Comment by ayan-biswas0412** [comment]
@Doringl @singhayushh @aniketsingh98571 can you comment on this is this thing needed ?

**Comment by aniketsingh98571** [comment]
> @Doringl @singhayushh @aniketsingh98571 can you comment on this is this thing needed ?

Hey @ayan-biswas0412 , I am not a good backend guy I think @singhayushh will be a good person to manage this.

**Comment by singhayushh** [comment]
If you are looking ahead to push this project as a live product, you can use these security packages to prevent DDoS attacks on the website.

**Pull request closed without merging**